Fix problem with applying pulse time filtering in LoadEventNexus

DannyHindson opened the pull request:
**Description of work.**

The WISH instrument scientists filter event nexus files by pulse time using some extra parameters on the LoadEventNexus algorithm:
```
LoadEventNexus(Filename=filename,OutputWorkspace=output,FilterByTimeStart=tmin,FilterByTimeStop=tmax,LoadMonitors='1',MonitorsAsEvents='1',FilterMonByTimeStart=tmin,FilterMonByTimeStop=tmax)
```
The pulse time filtering doesn't work if the input Nexus file isn't sorted on pulse time. A random subset of the events is returned. Note - I've recently fixed a similar problem in the FilterByTime algorithm (in https://github.com/mantidproject/mantid/pull/32499) but this PR fixes a separate bug in the LoadEventNexus code.

The previous approach of assuming you could convert the pulse time filter into an event id from\to range wasn't valid so I've just called the FilterByTime algorithm as a post process. This probably isn't as quick as the original method but don't think the original approach can be fixed.

There's another filter (FilterDuringPause) set up as a template method so I've done the same for the new FilterEventsByTime method This template approach possibly exists so the filter can work on the variety of workspace types that may need filtering (especially if the monitor data is filtered where it could be an event\matrix\group workspace or possibly even an EventWorkspaceCollection). Since FilterByTime doesn't have a feature to apply the filter in place I had to create a new method on EventWorkspaceCollection.

This fix has another positive side effect which is that if a FilterByTimeStart value is supplied that is larger than the duration of the run, no events are returned. The original code returned all the events in this case.

I noticed two other problems but haven't fixed because didn't seem as urgent:
- The parameters FilterMonByTimeStart and FilterMonByTimeStop don't seem to be implemented in the code ie they don't do anything. All the event mode files I've seen seem to have the monitor data captured in histogram mode anyway so maybe not a big deal
- I noticed that LoadEventNexus currently passes m_ws->getSingleHeldWorkspace() to filterDuringPause at one point instead of simply passing m_ws. I don't think this is correct because it only results in the first workspace (first period) being filtered but am leaving it because the pause filtering is not the subject of the bug reported

**To test:**

This load should return only 51098 events whereas before it returned over 11 million:
```
event_ws = Load(Filename=r'POLARIS00129560.nxs', FilterByTimeStart=308, FilterByTimeStop=309)
```

Fixes #32534.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Fix problem with applying pulse time filtering

The pulse time filtering previously assumed the events were sorted by
pulse time in the input Nexus file. This isn't always the case
The previous approach of assuming you could convert the pulse time
filter into an event id from\to range wasn't valid so I've just called
the FilterByTime algorithm as a post process. This probably isn't as
quick as the original method but don't think the original approach can be
fixed.
There's another filter (FilterDuringPause) set up as a template method
on the EventWorkspaceCollection class so I've done the same for the
FilterByTime method so it works on multi-period files. But since FilterByTime
doesn't have a feature to apply the filter in place I had to create a new
method on EventWorkspaceCollection
I noticed that LoadEventNexus currently passes m_ws->getSingleHeldWorkspace()
to filterDuringPause at one point instead of simply passing m_ws. I don't
think this is correct because it only results in the first workspace (period)
being filtered but am leaving it because not part of this bug fix

DannyHindson committed:
Fix failing doc test and add release note

DannyHindson committed:
Remove method name clash with filterByTime

martyngigg commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -1157,6 +1183,13 @@ void LoadEventNexus::loadEvents(API::Progress *const prog, const bool monitors)
+  if (is_time_filtered) {
+    // Now filter out the run and events, using the DateAndTime type.
+    // This will sort both by pulse time
+    m_ws->mutableRun().filterByTime(filter_time_start, filter_time_stop);
```
Does `FilterByTime` not do this itself?

DannyHindson replied:
Yes it does. I'd missed this. So this line can be removed I think

martyngigg replied:
Thanks. The unit test already checks that the logs have been filtered and that seems okay so I think we're good.

DannyHindson committed:
Remove redundant line to filter the sample logs

Run filtering already done in filterEventsByTime method via FilterByTime algorithm

martyngigg reviewed (approved):
I'm happy with this solution. The timings of either filtering in load or separately are not so much different and the benefit is we fix two bugs and consolidate code.

DannyHindson commented:
There's one piece of follow up I'd like to do on this - but suggest I do it separately.
One of the unit test files (V20_ESSIntegration_2018-12-13_0942.nxs) contains some unsorted events and I'd previously claimed this covered the 2 bugs that I've found and fixed. I've realised while working on this PR that it doesn't quite. Although the global event list is unsorted in this file, the event lists per spectrum are all sorted OK. So the unit test succeeds on the code before the bug fixes as well. I've asked Ron Smith if I can add one of the POLARIS nexus files into our test data set and am waiting for him to come back to me. I could create my own file I guess but since the other LoadEventNexusTest tests all use real files it seemed in keeping with that to add another real file from an instrument
